syntax = "proto3";
package backup_protocol;
import "google/protobuf/empty.proto";

service Replica{
    rpc Read(ReadDeleteRequest) returns (ReadResponse);
    rpc Write(WriteRequest) returns (WriteResponse);
    rpc Delete(ReadDeleteRequest) returns (Response);
    rpc HandleWrite(WriteRequest) returns (WriteResponse);
    rpc NotifyPrimary(ServerMessage) returns  (google.protobuf.Empty);
}

service RegistryServer{
    rpc RegisterReplica(ServerMessage) returns (ServerMessage);
    rpc GetReplicas(google.protobuf.Empty) returns (ServerListResponse);
}

message ServerMessage {
    string uuid = 1;
    string address = 2;
}

message ServerListResponse {
    repeated ServerMessage serverList = 1;
}

message WriteRequest {
    string name = 1;
    string content = 2;
    string uuid = 3;
}

message WriteResponse {
    Status status = 1;
    string uuid  = 2;
    string version = 3;
}

message ReadDeleteRequest{
    string uuid=1;
}

message ReadResponse{
    Status status=1;
    string name=2;
    string content=3;
    string version=4;
}

message Response{
    Status response=1;
}

message FileDetail{
    string file_name=1;
    string version=2;
}

message FileList{
    repeated FileDetail fileList = 1;
}


enum Status {
    SUCCESS = 0;
    FAIL = 1;
}